src/main/main.ts

```diff
@@ -188,214 +189,195 @@ function initializeApp() {
+
+  // TODO: 이어받기 되면 좋을 듯
+  const options: ElectronDLOptions = {
+    onStarted: (downloadItem: DownloadItem) => {
+      console.log("Starts to download:", downloadItem);
+    },
+    onProgress: (status: IDownloadProgress) => {
+      const percent = (status.percent * 100) | 0;
       console.log(
-        "The existing and new configuration files has been merged:",
-        config
+        `Downloading ${downloadUrl}: ${status.transferredBytes}/${status.totalBytes} (${percent}%)`
       );
+      win?.webContents.send("update download progress", status);
+    },
+    directory: app.getPath("temp"),
+  };
+  console.log("Starts to download:", downloadUrl);
+  let dl: DownloadItem | null | undefined;
+  try {
+    dl = await download(win, downloadUrl, options);
+  } catch (error) {
+    win?.webContents.send("go to error page", "download-binary-failed");
+    throw new DownloadBinaryFailedError(downloadUrl);
+  }
 
-      lockfile.unlockSync(lockfilePath);
-      console.log(
-        "Removed 'encounter different version' lockfile at ",
-        lockfilePath
+  win?.webContents.send("update download complete");
+  const dlFname = dl?.getFilename();
+  const dlPath = dl?.getSavePath();
+  console.log("Finished to download:", dlPath);
+
+  const extractPath =
+    process.platform == "darwin" // .app으로 실행한 건지 npm run dev로 실행한 건지도 확인해야 함
+      ? path.dirname(path.dirname(path.dirname(path.dirname(app.getAppPath()))))
+      : path.dirname(path.dirname(app.getAppPath()));
+  console.log("The 9C app installation path:", extractPath);
+
+  const appDirName = app.getAppPath();
+  // FIXME: "config.json" 이거 하드코딩하지 말아야 함
+  const configFileName = "config.json";
+
+  // 압축 해제하기 전에 기존 설정 꿍쳐둔다. 나중에 기존 설정 내용이랑 새 디폴트 값들이랑 합쳐야 함.
+  const configPath = path.join(appDirName, configFileName);
+  const bakConfig = JSON.parse(
+    await fs.promises.readFile(configPath, { encoding: "utf-8" })
+  );
+  console.log("The existing configuration:", bakConfig);
+
+  if (process.platform == "win32") {
+    // 윈도는 프로세스 떠 있는 실행 파일을 덮어씌우거나 지우지 못하므로 이름을 바꿔둬야 함.
+    const src = app.getPath("exe");
+    const basename = path.basename(src);
+    const dirname = path.dirname(src);
+    const dst = path.join(dirname, "bak_" + basename);
+    await fs.promises.rename(src, dst);
+    console.log("The executing file has renamed from", src, "to", dst);
+
+    // TODO: temp directory 앞에 9c-updater- 접두어
+    const tempDir = await tmpName();
+
+    // ZIP 압축 해제
+    console.log("Start to extract the zip archive", dlPath, "to", tempDir);
+
+    await extractZip(dlPath, {
+      dir: tempDir,
+      onEntry: (_, zipfile) => {
+        const progress = zipfile.entriesRead / zipfile.entryCount;
+        win?.webContents.send("update extract progress", progress);
+      },
+    });
+    win?.webContents.send("update extract complete");
+    console.log("The zip archive", dlPath, "has extracted to", tempDir);
+    win?.webContents.send("update copying progress");
```

Comment:
No need to check null.
```suggestion
    win.webContents.send("update extract complete");
    console.log("The zip archive", dlPath, "has extracted to", tempDir);
    win.webContents.send("update copying progress");
```